### ISSUE TITLE
More Testing on color match algorithm

### ISSUE THREAD

**Issue by dayuwater** [dayuwater/4244#10]
@poorvi-vora 

In this week I did more testing on my color matching algorithm. I tested my program against some real pictures, not the n*n color block pictures, and it works without changing any codes, even if the fragments are L shaped ( the complicated case where edge feature must be considered too)

Perhaps I should change the content of my presentation to reflect my most recent achievement. I think the color block case is still useful because it shows how my algorithm works in a more obvious way than the real picture case, but I plan to do a demo on my L-shaped color matching using real pictures. Do you think it is a good idea?

All of my previous works is on Emgu CV 2.4.9, but that is an old version. When I started my project, both OpenCV 3.0 and Emgu CV 3.0 just came out, and they use a way better algorithm that utilizes the GPU acceleration in a even better way. Because most of the tutorials did not use 3.0, and there are a lot of syntax change in 3.0 version, I chose not to use 3.0. Because I understand the algorithm and got a very good shape of my program now, it should be a good time for me to rewrite my program in Emgu CV 3.0 and clean up some redundant code , and I hope my program could run faster. http://www.emgu.com/wiki/index.php/Tutorial#Upgrading_from_Emgu_CV_2.x_to_3.x 
Pay special attention to the T-API feature, it says the program will run way faster if there is a discrete GPU in a computer, my computer has a good discrete GPU, so I could utilize this.

**Comment by poorvi-vora** [comment]
Tan:
1. but I plan to do a demo on my L-shaped color matching using real
   pictures. Do you think it is a good idea?
- Yes. Can you also show what the existing code (that is not yours) will do
  to it? Will the two be different? Please answer.
1. Emgu CV 3.0
- You can try this, but I am a little nervous if you plan to do a live
  demo, you should test it.

**Comment by dayuwater** [comment]
@poorvi-vora 
See my email I replied to you.

Here is their documentation on how to use their new data structure for images:
http://www.emgu.com/wiki/index.php/Working_with_Images

For all of the tests below, no manual memory allocation is used
I am doing some stress testing in order to justify whether Emgu CV 3.1 ( this is the newest version and the version I plan to use, but basically no algorithmic change compare to 3.0 ), here is the result:
1. Painting

1-1. Paint a 120*120 pixel picture with blue for 40000 times
3.1: 1.94,1.89,1.98,1.96,1.93 sec ( 40000 is about the crashing point, used over 2G of memory )
2.4: 5.36, 5.18, 5.04, 5.74, 4.80 sec ( crashes at 40000 times, so I test with 20000 times and double the result)
Summary: 3.1 is about 2.5 times faster than 2.4, and it supports more memory ( this is why 2.4 crashes so early ) . I found on their website that the data structure for images in 3.1 is significantly enhanced and supports automatic garbage collection.

1-2: Paint a 1200_1200 pixel picture with blue for 400 times ( again, this is the crash point)
3.1: 0.89, 0.84, 0.89, 0.91, 0.87 sec
2.4: 1, 0.96, 0.98, 0.9, 1.04 sec ( again, this crashes at 400 times, so I used the 2_200 method )
Summary: 3.1 only does slightly faster than 2.4 in this case, but 3.1 could handle 2 times the memory without crashing

1-3: Loading picture
3.1 loads picture way faster than 2.4 ( about 10 times), and it can supports 10 times the memory than 2.4.
1.  Image Manipulations ( This is essential for shifting fragments into the correct position, even tweaking algorithm can run a lot faster if this can be improved)
   2.4 Supports all of the basic image manipulations in their class, but I need write my own wrapper class in order to fully utilize the efficiency provided in 3.1 
   2-1: Rotate 120_120 pixel picture for 40000 times
   2-2: Rotate 1200_1200 pixel picture for 400 times
   2-3: Translate 120_120 pixel picture for 40000 times
   2-4: Translate 1200_1200 pixel picture for 400 times

The edge feature extraction only works in the 2.4 class, but there is a way to convert both ways.

As for the real part that do the actual feature detection, because both the edge matching and color matching detects the matching edge without using any feature in Emgu CV, so switching to 3.1 will not work in this part, but I can still improve the performance by cleaning up the code.
